File: src/main/java/io/github/welingtonmonteiro/multiplerun/LogFilter.java
package io.github.welingtonmonteiro.multiplerun;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * An advanced, composable filter for the aggregated Logs tab, kept free of IDE dependencies so it
 * is unit-testable in isolation. A line is shown when it passes, in order:
 * <ol>
 *   <li><b>App</b> - only lines from the selected application (or every app when none is set);</li>
 *   <li><b>Level</b> - only lines at or above a minimum log level, detected from the text;</li>
 *   <li><b>Terms</b> - several comma-separated terms combined with {@link Mode#ANY}/{@link Mode#ALL},
 *       matched as substrings or regular expressions, case-sensitive or not, and either kept
 *       (include) or dropped ({@link #exclude}).</li>
 * </ol>
 */
public final class LogFilter {

    /** Ordered by severity, so a minimum level is a simple ordinal comparison. */
    public enum Level { TRACE, DEBUG, INFO, WARN, ERROR }

    /** How several terms combine. */
    public enum Mode { ANY, ALL }

    private static final Pattern P_ERROR = Pattern.compile("\\b(ERROR|ERR|FATAL|SEVERE)\\b");
    private static final Pattern P_WARN = Pattern.compile("\\b(WARN|WARNING)\\b");
    private static final Pattern P_INFO = Pattern.compile("\\bINFO\\b");
    private static final Pattern P_DEBUG = Pattern.compile("\\bDEBUG\\b");
    private static final Pattern P_TRACE = Pattern.compile("\\bTRACE\\b");

    private final String app;            // null = every app
    private final List<String> terms;    // already trimmed, empties dropped
    private final Mode mode;
    private final boolean exclude;       // true = hide matching lines; false = show only matches
    private final boolean caseSensitive;
    private final boolean regex;
    private final Level minLevel;        // null = every level
    private final List<Pattern> patterns; // compiled regex per term (null slot = invalid regex)

    public LogFilter(String app, List<String> terms, Mode mode, boolean exclude,
                     boolean caseSensitive, boolean regex, Level minLevel) {
        this.app = app == null || app.trim().isEmpty() ? null : app;
        this.terms = terms == null ? new ArrayList<>() : terms;
        this.mode = mode == null ? Mode.ANY : mode;
        this.exclude = exclude;
        this.caseSensitive = caseSensitive;
        this.regex = regex;
        this.minLevel = minLevel;
        this.patterns = new ArrayList<>();
        if (regex) {
            final int flags = caseSensitive ? 0 : Pattern.CASE_INSENSITIVE;
            for (String term : this.terms) {
                try {
                    patterns.add(Pattern.compile(term, flags));
                } catch (PatternSyntaxException e) {
                    patterns.add(null); // an invalid regex simply never matches
                }
            }
        }
    }

    /** The identity filter: every line of every app passes. */
    public static LogFilter all() {
        return new LogFilter(null, new ArrayList<>(), Mode.ANY, false, false, false, null);
    }

    /**
     * Convenience factory. In substring mode {@code rawTerms} is split on commas into several terms
     * (combined by {@code mode}); in regex mode the whole string is a single pattern, since a regex
     * quantifier like {@code \d{3,}} legitimately contains a comma.
     */
    public static LogFilter from(String app, String rawTerms, Mode mode, boolean exclude,
                                 boolean caseSensitive, boolean regex, Level minLevel) {
        final List<String> terms;
        if (regex) {
            final String single = rawTerms == null ? "" : rawTerms.trim();
            terms = single.isEmpty() ? new ArrayList<>() : new ArrayList<>(java.util.Collections.singletonList(single));
        } else {
            terms = splitTerms(rawTerms);
        }
        return new LogFilter(app, terms, mode, exclude, caseSensitive, regex, minLevel);
    }

    /** Splits a comma-separated filter string into trimmed, non-empty terms. */
    public static List<String> splitTerms(String raw) {
        final List<String> out = new ArrayList<>();
        if (raw == null) {
            return out;
        }
        for (String part : raw.split(",")) {
            final String term = part.trim();
            if (!term.isEmpty()) {
                out.add(term);
            }
        }
        return out;
    }

    /** True when {@code line} from {@code lineApp} passes every enabled criterion. */
    public boolean accepts(String lineApp, String line) {
        if (app != null && !app.equals(lineApp)) {
            return false;
        }
        if (minLevel != null) {
            final Level level = detectLevel(line);
            if (level == null || level.ordinal() < minLevel.ordinal()) {
                return false;
            }
        }
        if (terms.isEmpty()) {
            return true; // nothing to include/exclude
        }
        final boolean combined = matchesTerms(line == null ? "" : line);
        return exclude ? !combined : combined;
    }

    private boolean matchesTerms(String line) {
        boolean any = false;
        boolean all = true;
        for (int i = 0; i < terms.size(); i++) {
            final boolean matched = termMatches(i, line);
            any |= matched;
            all &= matched;
        }
        return mode == Mode.ALL ? all : any;
    }

    private boolean termMatches(int index, String line) {
        if (regex) {
            final Pattern pattern = patterns.get(index);
            return pattern != null && pattern.matcher(line).find();
        }
        final String term = terms.get(index);
        return caseSensitive
                ? line.contains(term)
                : line.toLowerCase(Locale.ROOT).contains(term.toLowerCase(Locale.ROOT));
    }

    /** Best-effort log level of a line from common level tokens, or null when none is found. */
    public static Level detectLevel(String line) {
        if (line == null) {
            return null;
        }
        final String upper = line.toUpperCase(Locale.ROOT);
        if (P_ERROR.matcher(upper).find()) {
            return Level.ERROR;
        }
        if (P_WARN.matcher(upper).find()) {
            return Level.WARN;
        }
        if (P_INFO.matcher(upper).find()) {
            return Level.INFO;
        }
        if (P_DEBUG.matcher(upper).find()) {
            return Level.DEBUG;
        }
        if (P_TRACE.matcher(upper).find()) {
            return Level.TRACE;
        }
        return null;
    }
}

File: src/main/kotlin/io/github/welingtonmonteiro/lanes/LogFilter.kt
package io.github.welingtonmonteiro.lanes

import java.util.Locale
import java.util.regex.Pattern
import java.util.regex.PatternSyntaxException

/**
 * An advanced, composable filter for the aggregated Logs tab, kept free of IDE dependencies so it
 * is unit-testable in isolation. A line is shown when it passes, in order:
 *  1. **App** - only lines from the selected application (or every app when none is set);
 *  2. **Level** - only lines at or above a minimum log level, detected from the text;
 *  3. **Terms** - several comma-separated terms combined with [Mode.ANY]/[Mode.ALL], matched as
 *     substrings or regular expressions, case-sensitive or not, and either kept (include) or
 *     dropped ([exclude]).
 */
class LogFilter(
    app: String?,
    terms: List<String>?,
    mode: Mode?,
    private val exclude: Boolean,
    private val caseSensitive: Boolean,
    private val regex: Boolean,
    private val minLevel: Level?,
) {

    /** Ordered by severity, so a minimum level is a simple ordinal comparison. */
    enum class Level { TRACE, DEBUG, INFO, WARN, ERROR }

    /** How several terms combine. */
    enum class Mode { ANY, ALL }

    private val app: String? = if (app == null || app.trim().isEmpty()) null else app
    private val terms: List<String> = terms ?: emptyList()
    private val mode: Mode = mode ?: Mode.ANY

    /** Compiled regex per term (a null slot is an invalid regex, which never matches). */
    private val patterns: List<Pattern?> =
        if (regex) {
            val flags = if (caseSensitive) 0 else Pattern.CASE_INSENSITIVE
            this.terms.map { term ->
                try {
                    Pattern.compile(term, flags)
                } catch (e: PatternSyntaxException) {
                    null
                }
            }
        } else {
            emptyList()
        }

    /** True when [line] from [lineApp] passes every enabled criterion. */
    fun accepts(lineApp: String?, line: String?): Boolean {
        if (app != null && app != lineApp) {
            return false
        }
        if (minLevel != null) {
            val level = detectLevel(line)
            if (level == null || level.ordinal < minLevel.ordinal) {
                return false
            }
        }
        if (terms.isEmpty()) {
            return true // nothing to include/exclude
        }
        val combined = matchesTerms(line ?: "")
        return if (exclude) !combined else combined
    }

    private fun matchesTerms(line: String): Boolean {
        var any = false
        var all = true
        for (i in terms.indices) {
            val matched = termMatches(i, line)
            any = any || matched
            all = all && matched
        }
        return if (mode == Mode.ALL) all else any
    }

    private fun termMatches(index: Int, line: String): Boolean {
        if (regex) {
            val pattern = patterns[index]
            return pattern != null && pattern.matcher(line).find()
        }
        val term = terms[index]
        return if (caseSensitive) {
            line.contains(term)
        } else {
            line.lowercase(Locale.ROOT).contains(term.lowercase(Locale.ROOT))
        }
    }

    companion object {
        private val P_ERROR = Pattern.compile("\\b(ERROR|ERR|FATAL|SEVERE)\\b")
        private val P_WARN = Pattern.compile("\\b(WARN|WARNING)\\b")
        private val P_INFO = Pattern.compile("\\bINFO\\b")
        private val P_DEBUG = Pattern.compile("\\bDEBUG\\b")
        private val P_TRACE = Pattern.compile("\\bTRACE\\b")

        /** The identity filter: every line of every app passes. */
        @JvmStatic
        fun all(): LogFilter = LogFilter(null, emptyList(), Mode.ANY, false, false, false, null)

        /**
         * Convenience factory. In substring mode [rawTerms] is split on commas into several terms
         * (combined by [mode]); in regex mode the whole string is a single pattern, since a regex
         * quantifier like `\d{3,}` legitimately contains a comma.
         */
        @JvmStatic
        fun from(
            app: String?,
            rawTerms: String?,
            mode: Mode?,
            exclude: Boolean,
            caseSensitive: Boolean,
            regex: Boolean,
            minLevel: Level?,
        ): LogFilter {
            val terms: List<String> = if (regex) {
                val single = rawTerms?.trim() ?: ""
                if (single.isEmpty()) emptyList() else listOf(single)
            } else {
                splitTerms(rawTerms)
            }
            return LogFilter(app, terms, mode, exclude, caseSensitive, regex, minLevel)
        }

        /** Splits a comma-separated filter string into trimmed, non-empty terms. */
        @JvmStatic
        fun splitTerms(raw: String?): List<String> {
            val out = ArrayList<String>()
            if (raw == null) {
                return out
            }
            for (part in raw.split(",")) {
                val term = part.trim()
                if (term.isNotEmpty()) {
                    out.add(term)
                }
            }
            return out
        }

        /** Best-effort log level of a line from common level tokens, or null when none is found. */
        @JvmStatic
        fun detectLevel(line: String?): Level? {
            if (line == null) {
                return null
            }
            val upper = line.uppercase(Locale.ROOT)
            return when {
                P_ERROR.matcher(upper).find() -> Level.ERROR
                P_WARN.matcher(upper).find() -> Level.WARN
                P_INFO.matcher(upper).find() -> Level.INFO
                P_DEBUG.matcher(upper).find() -> Level.DEBUG
                P_TRACE.matcher(upper).find() -> Level.TRACE
                else -> null
            }
        }
    }
}
